presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java

```diff
@@ -68,6 +72,24 @@ public HivePageSourceProvider(
     @Override
     public ConnectorPageSource createPageSource(ConnectorSession session, ConnectorSplit split, List<ColumnHandle> columns)
     {
+        if (HiveSessionProperties.getReadAsQueryUser(session)) {
+            UserGroupInformation ugi = UgiUtils.getUgi(session.getUser());
+            try {
+                return ugi.doAs((PrivilegedExceptionAction<ConnectorPageSource>) () ->
```

Comment:
Is it sufficient to only `doAs` for the initial creation of the page source?  Historically, we have had problems with Haooop APIs and the multithreaded nature of Presto.  In a standard query this page source will be accessed from many threads, so if in the middle of the query the thread preforms a new "read", the UGI will not be set on the thread.